internal/state/store.go

```diff
@@ -932,3 +911,49 @@ func int64ToBytes(i int64) []byte {
 	n := binary.PutVarint(buf, i)
 	return buf[:n]
 }
+
+//---------------------------------- KEY ENCODING -----------------------------------------
+
+const (
+	// subkeys must be unique within a single DB
+	subkeyValidators      = int64(5)
+	subkeyConsensusParams = int64(6)
+	subkeyABCIResponses   = int64(7)
+)
+
+var (
+	stateKey = []byte("stateKey")
+
+	lastABCIResponseKey              = []byte("lastABCIResponseKey")
+	lastABCIResponsesRetainHeightKey = []byte("lastABCIResponsesRetainHeight")
+	offlineStateSyncHeight           = []byte("offlineStateSyncHeightKey")
+)
+
+func encodeKey(height, prefix int64) []byte {
+	res, err := orderedcode.Append(nil, height, prefix)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+func validatorsKey(height int64) []byte {
+	// Since CometBFT requests block H and validators H+1 on every height, we
```

Comment:
This is the most controversial change. But this is undoubtedly an improvement because the access pattern is clear (block H, parts H, validators H+1). Any thoughts?

Reply:
I think it should be fine although there are places where we fetch the validators info explicitly (like RPC `status` and pruning). But those are less frequent than block validation. Another frequent access pattern is loading the lastCommit (validators from height -1 ) which is called on every `ApplyBlock`